dir.c

```diff
@@ -1041,8 +1041,55 @@ dir_chdir0(VALUE path)
         rb_sys_fail_path(path);
 }
 
-static int chdir_blocking = 0;
-static VALUE chdir_thread = Qnil;
+static struct {
+    VALUE thread;
+    VALUE path;
+    int line;
+    int blocking;
+} chdir_lock = {
+    .blocking = 0, .thread = Qnil,
+    .path = Qnil, .line = 0,
+};
+
+static void
+chdir_enter(void)
+{
+    if (chdir_lock.blocking == 0) {
+        chdir_lock.path = rb_source_location(&chdir_lock.line);
+    }
+    chdir_lock.blocking++;
+    if (NIL_P(chdir_lock.thread)) {
+        chdir_lock.thread = rb_thread_current();
+    }
+}
+
+static void
+chdir_leave(void)
+{
+    chdir_lock.blocking--;
+    if (chdir_lock.blocking == 0) {
+        chdir_lock.thread = Qnil;
+        chdir_lock.path = Qnil;
+        chdir_lock.line = 0;
+    }
+}
+
+static int
+chdir_alone_block_p(void)
+{
+    int block_given = rb_block_given_p();
+    if (chdir_lock.blocking > 0) {
+        if (rb_thread_current() != chdir_lock.thread)
+            rb_raise(rb_eRuntimeError, "conflicting chdir during another chdir block");
+        if (!block_given) {
+            rb_warn("conflicting chdir during another chdir block");
+            if (!NIL_P(chdir_lock.path)) {
+                rb_compile_warn(RSTRING_PTR(chdir_lock.path), chdir_lock.line, "here");
+            }
```

Comment:
Would it be possible to make it a single warning?

Warnings that are split in two are a bit painful for libraries and applications that hook into `Warning.warn`. e.g. https://github.com/jeremyevans/ruby-warning or https://github.com/Shopify/deprecation_toolkit

Reply:
It is for editors to jump to each lines.

Reply:
Sorry I wasn't clear. I didn't mean not to include both locations, but to include them both in a single warning / string.

Reply:
e.g.:

```ruby
rb_warn("conflicting chdir during another chdir block\nPrevious chdir was here: #{location}")
```

Instead of 

```ruby
rb_warn("conflicting chdir during another chdir block")
rb_warn("Previous chdir was here: #{location}")
```

Otherwise we need to do things like this: https://github.com/Shopify/deprecation_toolkit/blob/main/lib/deprecation_toolkit/warning.rb#L9-L23

Reply:
I wanted to jump to both inner `chdir` line and caller `chdir` block line.
To using the messages for tag jump, each needs to be separate lines.

Reply:
What I'm trying to say it that it can be two lines, and look identical to two `rb_warn` calls, but be a single `rb_warn` call (with a multiline string).

```ruby
rb_warn(
  "conflicting chdir during another chdir block\n"
  "#{location}: warning: previous chdir was here"
);
```

That would work with your editor, no?

But I can also try to make this work later if you want.

Reply:
I think I fixed it.
Is it OK?

Reply:
Looks great. Thank you very much!